### PR TITLE
Enable docker daemon cross-building for ARM and i386

### PR DESCRIPTION
**Pull request by lifeeth**
This commit does the following:
1. Separates out the cross-building parts of env Dockerfile into Dockerfile.cross
2. Enables daemon building for ARM and i386 on Linux.

Signed-off-by: Praneeth Bodduluri lifeeth@gmail.com

### CODE REVIEW

**Comment by cpuguy83** [comment]
cc @tianon

**Comment by brahmaroutu** [comment]
@lifeeth We are approaching the problem slightly differently. 
It is going to be hard to keep multiple Dockerfiles up-to-date. I am in the process of issuing a PR that uses bash processing on the Dockerfile to create a buildable Dockerfile  Please see my local branch
https://github.com/brahmaroutu/docker/tree/dockerfile_gen 
If you look at PR #11298 which separates a script that is runnable using just make.sh, this assumes that you have all requirements on the host. 
With a script and the auto-generated Dockerfile I believe we can achieve this PR?

File: project/make/cross [hunk 1]
```diff
@@ -28,6 +30,21 @@ for platform in $DOCKER_CROSSPLATFORMS; do
 			export LDFLAGS_STATIC_DOCKER="" # we just need a simple client for these platforms
 			export BUILDFLAGS=( "${ORIG_BUILDFLAGS[@]/ daemon/}" ) # remove the "daemon" build tag from platforms that aren't supported
 		fi
+		if [ $GOARCH = "arm" ] && [ $GOOS = "linux" ]
+		then
```
Comment (WarheadsSE) [inline project/make/cross]: What are the default CXXFLAGS on this toolchain?

File: Dockerfile.cross [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	&& make install
+
+# Compile Go for cross compilation
+ENV DOCKER_CROSSPLATFORMS \
+	linux/386 linux/arm \
+	darwin/amd64 darwin/386 \
+	freebsd/amd64 freebsd/386 freebsd/arm \
+	windows/amd64 windows/386
+
+# (set an explicit GOARM of 5 for maximum compatibility)
```
Comment (WarheadsSE) [inline Dockerfile.cross]: While I +10 for compatibility, I want to make note of the performance hit versus newer versions and floating point support. The instructions are fully compatible, but soft-float binaries in hard-float userlands MUST be statically compiled. (Which this is)

**Comment by lifeeth** [comment]
@brahmaroutu This method of splitting the Dockerfile was suggested by the Docker team - If your PR is the preferred way by the team - I could redo this one to fit into your method.

File: project/make/cross (hunk 1 above)
Comment (lifeeth) [inline project/make/cross]: @WarheadsSE If you mean for arm-linux-gnueabi-gcc they are the defaults from https://wiki.ubuntu.com/ToolChain/CompilerFlags

CGO itself has no default CXXFLAGS.

**Comment by brahmaroutu** [comment]
@lifeeth I will let @tianon comment further. I would like to see on solution for all.

**Comment by tianon** [comment]
Yeah, I'm definitely not a fan of how much overlap there is between the `Dockerfile`s because that increases the maintenance burden.

Right now, you can build a working Docker daemon on an ARM box natively with `./hack/make.sh dynbinary`.  What we don't have are officially released "daemon binaries" for ARM, which is where we'd need either the `Dockerfile` to be able to cross-compile them (as this PR is working towards), or we need to revamp our release process to have dedicated "release build" hardware, which further complicates an already lengthy release process, but will likely be necessary at some point (or at least some sort of virtualization for "natively" building these binaries for the target architectures).

File: project/make/cross [hunk 3]
```diff
@@ -6,7 +6,9 @@ DEST=$1
 # explicit list of os/arch combos that support being a daemon
 declare -A daemonSupporting
 daemonSupporting=(
-	[linux/amd64]=1
+	[linux/amd64]=1,
+	[linux/arm]=1,
```
Comment (tianon) [inline project/make/cross]: As a side note, these commas are superfluous.

File: project/make/cross [hunk 4]
```diff
@@ -28,6 +30,21 @@ for platform in $DOCKER_CROSSPLATFORMS; do
 			export LDFLAGS_STATIC_DOCKER="" # we just need a simple client for these platforms
 			export BUILDFLAGS=( "${ORIG_BUILDFLAGS[@]/ daemon/}" ) # remove the "daemon" build tag from platforms that aren't supported
 		fi
+		if [ $GOARCH = "arm" ] && [ $GOOS = "linux" ]
+		then
+			export CC=arm-linux-gnueabi-gcc
+			export LD_FLAGS="-extld=arm-linux-gnueabi-gcc"
+			export CGO_ENABLED=1
+			export LIBRARY_PATH=$LIBRARY_PATH:/usr/arm-linux-gnueabi/lib
```
Comment (tianon) [inline project/make/cross]: I'm pretty sure these will cause problems for users who are compiling outside the build containers, as I described in my earlier comment on this PR (`./hack/make.sh dynbinary` natively on the target architecture).

Comment (lifeeth) [inline project/make/cross]: Agreed @tianon - Didn't think of this - How about we create a new make target like so 
```
make cross-build
```
and move the changes which cross-compile there?

**Comment by lifeeth** [comment]
@tianon I have made some changes to fix the case when the the hack/make.sh cross is run natively. I tried to ensure that there are no overlaps between Dockerfile and Dockerfile.cross.

**Comment by ehazlett** [comment]
ping @tianon

**Comment by alexandrosm** [comment]
@tianon is there a way we could make it work for non-linux and better integrate with your release process, or are you effectively saying that anyone from outside the docker team cannot address this issue?

**Comment by tianon** [comment]
I'm saying that someone needs to come up with a release process that both
solves this problem and isn't too complicated for the release maintainer
(ie, stuff like manually SSHing into build machines).  Definitely doesn't
have to be someone on "the Docker team", but they're traditionally the most
well-versed with our current release process (documented in
"project/RELEASE-CHECKLIST.md").

**Comment by cpuguy83** [comment]
@tianon Should we go ahead and close this one?

**Comment by bfirsh** [comment]
@tianon Why does this PR not give us cross-compiled binaries for releases? Is there a change to how the release process would work due to this?

**Comment by cpuguy83** [comment]
Closing this one for now since it's not mergable.
There is work to be done on our release process to be able to handle adding support for other archs.

@lifeeth thanks again for the contribution

**Pull request closed without merging**